File: httpserve/permission.go
package httpserve

import (
	"encoding/json"
	"os"
	"strings"
	"sync"
)

// Permissions is the black/white list gating data + API commands, keyed by
// "COMMAND::collection". It mirrors doptime's single-list on/off model
// (command::key::on/off) so existing operational habits carry over; here the
// "key" is a collection name (or an API endpoint name).
//
// Default is DENY: a pair that was never granted is refused. (The former
// dev-only AutoAuth grant-on-first-use was removed — grants are always explicit.)
//
// This in-memory implementation is sufficient for a single process; persist it
// with SaveJSON / LoadJSON, or for a cluster back the same calls with a shared
// dopdb collection — the HTTP layer only calls Allowed/Grant/Deny.
type Permissions struct {
	mu sync.RWMutex
	m  map[string]bool // "CMD::coll" -> on
}

// NewPermissions returns an empty permission set (default deny).
func NewPermissions() *Permissions {
	return &Permissions{m: map[string]bool{}}
}

func permKey(cmd, coll string) string {
	return strings.ToUpper(cmd) + "::" + coll
}

// Allowed reports whether (cmd, coll) is explicitly permitted. Unknown pairs are
// denied.
func (p *Permissions) Allowed(cmd, coll string) bool {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.m[permKey(cmd, coll)]
}

// Grant adds an allow entry.
func (p *Permissions) Grant(cmd, coll string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.m[permKey(cmd, coll)] = true
}

// Deny adds an explicit deny entry.
func (p *Permissions) Deny(cmd, coll string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.m[permKey(cmd, coll)] = false
}

// SaveJSON serialises the current permission map to a JSON file.
func (p *Permissions) SaveJSON(path string) error {
	p.mu.RLock()
	data, err := json.MarshalIndent(p.m, "", "  ")
	p.mu.RUnlock()
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

// LoadJSON deserialises a JSON file back into a Permissions instance.
func LoadJSON(path string) (*Permissions, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var m map[string]bool
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, err
	}
	return &Permissions{m: m}, nil
}

File: ts/src/permission.ts
// permission.ts — the command/collection permission gate, equivalent to Go's
// httpserve.Permissions. Keyed by "COMMAND::collection". Default is DENY: a pair
// that was never granted is refused. There is no AutoAuth (grant-on-first-use);
// grants are always explicit.
//
//   const perms = new Permissions();
//   perms.grant("HGET", "User").grant("HSET", "User");
//   serve({ schema, mongo, jwtSecret, permissions: perms });

import { readFile, writeFile } from "node:fs/promises";

export class Permissions {
  private m = new Map<string, boolean>();

  private static key(cmd: string, coll: string): string {
    return `${cmd.toUpperCase()}::${coll}`;
  }

  /** Reports whether (cmd, coll) is explicitly permitted. Unknown pairs → false. */
  allowed(cmd: string, coll: string): boolean {
    return this.m.get(Permissions.key(cmd, coll)) === true;
  }

  /** Add an allow entry. Chainable. */
  grant(cmd: string, coll: string): this {
    this.m.set(Permissions.key(cmd, coll), true);
    return this;
  }

  /** Add an explicit deny entry (beats nothing in particular; absence already
   * denies — kept for parity with the Go API and to override a prior grant). */
  deny(cmd: string, coll: string): this {
    this.m.set(Permissions.key(cmd, coll), false);
    return this;
  }

  /** Plain-object view of the permission map (for serialization/inspection). */
  toJSON(): Record<string, boolean> {
    return Object.fromEntries(this.m);
  }

  /** Persist the permission map to a JSON file. */
  async saveJSON(path: string): Promise<void> {
    await writeFile(path, JSON.stringify(this.toJSON(), null, 2), "utf8");
  }

  /** Load a permission map previously written with saveJSON. */
  static async loadJSON(path: string): Promise<Permissions> {
    const data = await readFile(path, "utf8");
    const obj = JSON.parse(data) as Record<string, boolean>;
    const p = new Permissions();
    for (const [k, v] of Object.entries(obj)) p.m.set(k, v === true);
    return p;
  }
}
